File: week_8_and_9/4_Ruby/rectangle_class/my_solution.rb
# U3.W8-9: Implement a Rectangle Class


# I worked on this challenge [by myself, with: ].

# 2. Pseudocode
# 
# class rectangle with instance methods 
# area
# perimeter
# diagonal
# and square?(boolean)
# 
# CLASS Rectangle
# 	
# 	METHOD initialize which takes 2 arguments
# 		set @width = width
# 		set @height = height
# 	END METHOD
# 	
# 	METHOD area
# 		return length * height
# 	END METHOD
# 	
# 	METHOD perimeter
# 		return (length + height) * 2
# 	END METHOD
# 	
# 	METHOD diagonal
# 	 return - use math to find the square root of (length**2 + height**2)
# 	END METHOD
# 	
# 	METHOD square?
# 	 return true if length == width
# 	 return false
# 	END METHOD
# 
# END CLASS




# 3. Initial Solution
class Rectangle
  attr_accessor :width, :height

  def initialize(width, height)
    @width  = width
    @height = height
  end

  # def ==(other)
  #   (other.width  == self.width && other.height == self.height ) ||
  #   (other.height == self.width && other.width  == self.height )
  # end
  
  def area
  	@width * @height
  end

	def perimeter
		(@width + @height) * 2
	end

	def diagonal
		Math.sqrt(@width**2 + @height**2)
	end

	def square?
		return true if @width == @height
		false
	end

end




# 4. Refactored Solution


# No refactor currently :+(
# Perhaps I will have time to come back over the weekend. But, doubtfull. 



# 1. DRIVER TESTS GO BELOW THIS LINE

square = Rectangle.new(20,20)
rectangle = Rectangle.new(10,20)

p square.area == 400
p square.perimeter == 80
p square.diagonal == 28.284271247461902
p square.square? == true


p rectangle.area == 200
p rectangle.perimeter == 60
p rectangle.diagonal == 22.360679774997898
p rectangle.square? == false


# 5. Reflection 

# What parts of your strategy worked? What problems did you face?
# 
# All of it really, after transfering the rspec to driver code writing the pseudo 
# was super easy. Pseudo to initial was like filling in the blanks.
# 
# What questions did you have while coding? What resources did you find to help you answer them?
# 
# How do I find the diagonal of a rectangle. I used a new site called google to find the answer.
# 
# What concepts are you having trouble with, or did you just figure something out? If so, what?
# 
# None
# 
# Did you learn any new skills or tricks?
# 
# Math.sqrt
# 
# How confident are you with each of the Learning Competencies?
# 
# Very
# 
# Which parts of the challenge did you enjoy?
# 
# Another quickie challenge which is so super great for me. 
# Really getting down to the wire on my list of stuff to do Pre-DBC
# 
# Which parts of the challenge did you find tedious?
# 
# I see your commented out method skeleton there. I wish I had time to 
# refactor my solution and learn how to use self like you are indicating.
# Perhaps I will have a chance this weekend
# Until then No refactor :+(

File: week_8_and_9/3_JavaScript/translate_to_js/my_solution.js
// ruby code to port

// class GuessingGame

//   def initialize(answer)
//     @answer = answer
//     @solved = false
//   end

//   def guess(guess)
//      @solved = false
//      if guess < @answer
//          :low
//      elsif guess > @answer
//          :high
//      elsif guess == @answer
//          @solved = true
//          :correct
//      end
//   end

//   def solved?
//      @solved
//   end

// end

// game = GuessingGame.new(10)

// p game.solved? == false
// p game.guess(9) == :low
// p game.solved? == false
// p game.guess(11) == :high
// p game.solved? == false
// p game.guess(10) == :correct
// p game.solved? == true



// U3.W8-9: Challenge you're converting: U2.W5: Build a simple guessing game


// I worked on this challenge [by myself, with: Allison Wong & Hunter Chpaman].

// 2. Pseudocode
// create a quessing game with 2 functions
// one to check if the game has been solved
// one to handle the actual solving of the game
// symbols will be subed in as strings

// VARIABLES
// var solved = false
// var secret_num = 10


// FUNCTION guess
// guess which takes a single int as an input
// IF LOOP
//  less than input return :LOW
//  greater than input return :HIGH
//  equal to retur :CORRECT && set solved to true
// END IF LOOP
// END FUNCTION

// FUNCTION solved?
// solved? takes no arguements
// returns the value of var solved
// END FUNCTION



// 3. Initial Solution
// var solved = false;
// var secret_number = 10;

// var guess = function(int) {
//     if (int < secret_number){
//         return "low";     
//     }
//     else if (int > secret_number) {
//         return "high";
//     }
//     else if (int === secret_number) {
//         solved = true;
//         return "correct";
//     }
// };

// var is_solved = function(){
//     return solved;
// };



// 4. Refactored Solution
var solved = false;
var secret_number = 0;

var set_number = function(int) {
  secret_number = int;
};

var guess = function(int) {
    if (int < secret_number){
        return "low";
    }
    else if (int > secret_number) {
        return "high";
    }
    else if (int === secret_number) {
        solved = true;
        return "correct";
    }
};

var is_solved = function(){
    return solved;
};





// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE
console.log(secret_number === 0);
set_number(10);
console.log(secret_number === 10);
console.log(guess(11) === "high");
console.log(guess(9) === "low");
console.log(is_solved() === false);
console.log(guess(10) === "correct");
console.log(is_solved() === true);




// 5. Reflection 

// What parts of your strategy worked? What problems did you face?
// 
// The strategy here was to pick the easiest possible ruby challenge and port it over to JS. Why go the easy route. Cause I'm
// lazy and so ready to be done with these so I can enjoy the city for a day before DBC starts. I faced no problems. 
// 
// What questions did you have while coding? What resources did you find to help you answer them?
// 
// Would I get in trouble for skipping the reflections.... :+)
// 
// What concepts are you having trouble with, or did you just figure something out? If so, what?
// 
// javascript syntax is def not commit to memory just yet. still find myself referencing quite a bit. 
// 
// Did you learn any new skills or tricks?
// 
// no
// 
// How confident are you with each of the Learning Competencies?
// 
// mildly confident
// 
// Which parts of the challenge did you enjoy?
// 
// pairing is always better than not.
// 
// Which parts of the challenge did you find tedious?
// 
// driver code in js is not something I understand at all. console.log statements it it then. 
